trl/trainer/grpo_trainer.py

```diff
@@ -1837,6 +1837,11 @@ def _compute_loss(self, model, inputs):
         elif self.importance_sampling_level == "sequence":
             log_importance_weights = (log_ratio * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)
             log_importance_weights = log_importance_weights.unsqueeze(-1)
+        elif self.importance_sampling_level == 'sequence_token':
+            # GSPO-token: sg[si(θ)] * πθ(yi,t)/sg[πθ(yi,t)]
+            seq_level_log_weight = (log_ratio * completion_mask).sum(-1) / completion_mask.sum(-1).clamp(min=1.0)
+            seq_level_log_weight = seq_level_log_weight.unsqueeze(-1).detach()  # Stop gradient
+            log_importance_weights = seq_level_log_weight + per_token_logps - per_token_logps.detach()
```

Comment:
```suggestion
            log_importance_weights = per_token_logps - per_token_logps.detach() + seq_level_log_weight
```
reorder to avoid precision errors

Reply:
wasn't referring to the dtype, just better to perform the subtraction on `per_token_logps - per_token_logps.detach()` first, to avoid catastrophic cancellation

Reply:
that make sense, fixed